### PR TITLE
#140 Give Information of Non-Automated Rules

### LINKED ISSUE #140: Give information about not-automated rules
As user of Zalando's API Linter I'd like to know which rules are not yet automated. This would improve API design process because as an API provider I would know which rules I should read. Furthermore it would help API reviewers to know on which aspects to concentrate because they are not covered by Zally.

Server should deliver in api-violations response a simple description text about not yet automated rules. This text should be configurable. CLI should display this text as part of summary.

### PR DESCRIPTION
**Pull request by netme**
Closes #140

### CODE REVIEW

File: server/src/main/resources/api/zally-api.yaml [hunk 1]
```diff
@@ -72,6 +72,8 @@ definitions:
         type: array
         items:
           $ref: '#/definitions/Violation'
+      message:
+        type: string
```
Comment (netme) [inline server/src/main/resources/api/zally-api.yaml]: @fmueller , I will need your quick feedback on API change

File: server/src/main/java/de/zalando/zally/ApiViolationsController.java [hunk 2]
```diff
@@ -19,6 +20,9 @@
     private final RulesValidator rulesValidator;
     private final ObjectMapper mapper;
 
+    @Value("${zally.message}")
+    private String message;
```
Comment (fmueller) [inline server/src/main/java/de/zalando/zally/ApiViolationsController.java]: Can you transform this into constructor injection like the other parameters? @Value does also work on constructor argument.

File: server/src/main/resources/api/zally-api.yaml (hunk 1 above)
Comment (fmueller) [inline server/src/main/resources/api/zally-api.yaml]: LGTM

A simple string field can be used for any type of message. I like it.

Comment (fmueller) [inline server/src/main/resources/api/zally-api.yaml]: It should be optional btw. So, server can response without it if value is not specified or empty.

File: server/src/test/java/de/zalando/zally/RestApiTest.java [hunk 3]
```diff
@@ -80,6 +80,9 @@ public void shouldValidateGivenApiDefinition() throws IOException {
         JsonNode violations = rootObject.get("violations");
         assertThat(violations).hasSize(1);
         assertThat(violations.get(0).get("title").asText()).isEqualTo("dummy1");
+
+        String message = rootObject.get("message").asText();
+        assertThat(message).isEqualTo("Test message");
```
Comment (fmueller) [inline server/src/test/java/de/zalando/zally/RestApiTest.java]: Can you also test message not specified in configuration?

File: server/src/main/resources/api/zally-api.yaml (hunk 1 above)
Comment (netme) [inline server/src/main/resources/api/zally-api.yaml]: 👍 Makes sense, will add it to spec

File: server/src/main/java/de/zalando/zally/ApiViolationsController.java (hunk 2 above)
Comment (netme) [inline server/src/main/java/de/zalando/zally/ApiViolationsController.java]: I will take a look on it. Still a bit new to Spring.

File: server/src/test/java/de/zalando/zally/RestApiTest.java (hunk 3 above)
Comment (netme) [inline server/src/test/java/de/zalando/zally/RestApiTest.java]: I will take a look on it

File: server/src/main/java/de/zalando/zally/ApiViolationsController.java (hunk 2 above)
Comment (fmueller) [inline server/src/main/java/de/zalando/zally/ApiViolationsController.java]: No worries. :) As a rule in Java: always favor constructor based injection. It makes testing your code easier. If you only have constructor based injection, you don't need to start spring application context in your test and can simple mock your dependencies.

Comment (netme) [inline server/src/main/java/de/zalando/zally/ApiViolationsController.java]: Oh, thanks for explanation. I'm sure I will do the same trick in any other language, makes complete sense :-)

File: server/src/test/java/de/zalando/zally/RestApiTest.java (hunk 3 above)
Comment (netme) [inline server/src/test/java/de/zalando/zally/RestApiTest.java]: @fmueller , I believe I did a small improvement in this direction here: https://github.com/zalando-incubator/zally/pull/149/commits/b52b452ad22dfbdd3a6af0adba9ae74e0216155b#diff-681343b113913a3272e8cd1141af063bR38

But most probably it's still not that you expected. Should I create another `SpringBootTest` for that?

File: server/src/main/resources/api/zally-api.yaml (hunk 1 above)
Comment (netme) [inline server/src/main/resources/api/zally-api.yaml]: Fixed

File: server/src/test/java/de/zalando/zally/RestApiTest.java (hunk 3 above)
Comment (fmueller) [inline server/src/test/java/de/zalando/zally/RestApiTest.java]: My point is, you are only testing for "property has certain value and this value is returned in message field in json payload". What happens if this value is not given in the configuration (another test case) and what happens if the value is empty (also another test case).

Comment (netme) [inline server/src/test/java/de/zalando/zally/RestApiTest.java]: OK, I've got your point, will try to cover it.

File: cli/src/main/java/de/zalando/zally/cli/ResultPrinter.java [hunk 4]
```diff
@@ -32,6 +32,14 @@ public ResultPrinter(OutputStream outputStream) {
         writer = new OutputStreamWriter(outputStream);
     }
 
+    public void printMessage(String message) throws IOException {
+        if (!message.isEmpty()) {
+            printHeader(ANSI_WHITE, "Server message");
```
Comment (fmueller) [inline cli/src/main/java/de/zalando/zally/cli/ResultPrinter.java]: Is ANSI_WHITE actually readable in non-dark terminals?

Comment (netme) [inline cli/src/main/java/de/zalando/zally/cli/ResultPrinter.java]: Good point, I need to try that

**Comment by fmueller** [comment]
👍

**Comment by netme** [comment]
@fmueller resolving conflicts now :-)

**Comment by fmueller** [comment]
@netme I see, good luck. ^^ Will give you thumbs up afterwards :)

**Comment by netme** [comment]
Seems to be ready, let's see if Travis / Codacy will agree :-)

**Comment by fmueller** [comment]
👍

File: cli/src/test/java/de/zalando/zally/cli/ResultPrinterTest.java [hunk 5]
```diff
@@ -75,7 +75,7 @@ public void printsProperSummary() throws IOException {
         outContent.reset();
 
         resultPrinter.printSummary(Linter.violationTypes);
-        String expectedResult = resultPrinter.ANSI_WHITE + "\nSummary:\n"
+        String expectedResult = resultPrinter.ANSI_CYAN + "\nSummary:\n"
```
Comment (netme) [inline cli/src/test/java/de/zalando/zally/cli/ResultPrinterTest.java]: I also found out that we were using `ANSI_WHITE` for `Summary:` section and it was looking ugly on the white background, so good spot, @fmueller :-)

Comment (fmueller) [inline cli/src/test/java/de/zalando/zally/cli/ResultPrinterTest.java]: Nice finding :)
